content/blog/2019-12-21-logical-indexing-on-HB/index.md

```diff
@@ -0,0 +1,249 @@
+  stencil(&a_tile, &b_tile);
+```
+
+The **cached implementation** defines the same tensors, but additionally performs a caching routine that moves data from tensor `a_tile` to a new tensor `a_scr`, as follows:
+```c
+...
+tensor a_scr =
+  { .arr = scr,
+    .parent = &a_tile,
+    .n = tile_n+2, .m = tile_m+2,
+    .stride = m+2,
+    .phys_row_off = 1, .phys_col_off = 1,
+    .cached = true};
+
+  cache(&a_tile, &a_scr);
+  stencil(&a_scr, &b_tile);
+```
+
+The caching strategy reads a window around the tile data consisting of two additional rows and columns:
+```c
+void cache(tensor *input, tensor *cache) {
+  for (int x = -1; x < input->n+1; x++) {
+    for (int y = -1; y < input->m+1; y++) {
+      if(0 < log_to_phys_row(input, x) &&
+          log_to_phys_row(input, x) < top_level_tensor(input)->n &&
+         0 < log_to_phys_col(input, y) &&
+          log_to_phys_col(input, y) < top_level_tensor(input)->m ) {
+        log_write(cache, x, y,
+            log_read(input, x, y));
+      }
+    }
+  }
+```
+
+This caching routine shows how we can use the physical offset to arbitrarily shift the otherwise zero-based logical indexing.
+Using physical offset for this purpose is also cheating, since logical indexing may in principle be changed while preserving physical indexing.
+
+
+# Evaluation: Cached vs Non-cached
+Both stencil implementations ran on a gem5 simulation of a 4x4 HammerBlade architecture.
+Both stencil implementations were tested using matrices with elements equal to their row-major order indices. 
+This choice of matrix allowed for simple confirmation of correctness, since these matrices are not transformed by the stencil algorithm.
+
+We compare performance of both implementations by reading the `numCycles` statistic produced by the simulation for each input size and averaging the result over the four cores, and over three trials:
+
+**non-cached**
+| 4x4     | 8x8     | 16x16   |   34x34 |
+|---------|---------|---------|---------|
+|  7605   | 28740   |  123769 |  580502 |
+|         |         |         |         |
+
+**cached**
+| 4x4     | 8x8     | 16x16   | 34 x 34 |
+|---------|---------|---------|---------|
+|  14592  | 35961   |  115200 | 474928  |
+|         |         |         |         |
+
+# Discussion
+These results demonstrate that the cost of caching is amortized with the size of the input.
+
+Although this result is expected, I wondered why the cost of caching seemed so high.
+Cursory testing revealed that removing the branching from the cache loop removed ~10k cycles from the 4x4 and 8x8, ~40k cycles from the 16x16 input, and ~130k cycles from 34x34.
+These results suggest that a more efficient caching routine that minimizes branching would improve performance.
+
+Further analysis can be done to determine how much the simulation penalizes global memory vs scratchpad operations.
+This analysis would also inform whether caching on the output might present an opportunity for additional performance benefits.
```

Comment:
In fact, this might be a good case for execution on actual HammerBlade hardware or a simulator to get even more realistic numbers…

Reply:
Not sure what you mean here: it *was* executed on a simulator.
Do you mean "as opposed to" a simulator?

Reply:
Oops, sorry, I meant the “RTL simulator,” which is also a simulator, yes, but a higher-fidelity one then gem5.